assets/src/components/form/select/model-select.js

```diff
@@ -0,0 +1,170 @@
+/**
+ * External imports
+ */
+import Select from 'react-select';
+import { Component, Fragment } from '@wordpress/element';
+import { isEmpty, uniqueId, find, isUndefined } from 'lodash';
+import PropTypes from 'prop-types';
+
+/**
+ * WP dependencies
+ */
+import { withSelect } from '@wordpress/data';
+
+/**
+ * Internal imports
+ */
+import buildOptions from './build-options';
+import { MODEL_NAMES } from '../../../data/model';
+import {
+	REACT_SELECT_DEFAULTS,
+	REACT_SELECT_TYPES,
+} from './default-select-configuration';
+
+/**
+ * ModelSelect component.
+ * This is a component that will generate a react-select input for a given
+ * model and its entities (provided via props).
+ *
+ * @see https://deploy-preview-2289--react-select.netlify.com/props#prop-types
+ *   for options that can be passed through via the selectConfiguration prop.
+ *
+ * @param { Object } selectConfiguration  An object containing options for the
+ *                                          react-select component.
+ * @param { Array } modelEntities          Array of model entities
+ * @param { string } modelName              The name of the Model the entities
+ *                                          belong to.
+ * @param { function } mapOptionsCallback  This function will receive by
+ *   default the modelEntities, the modelName (and any custom Map provided) and
+ *   is expected to return an array of options to be used for the react-select
+ *   component.
+ * @param { Object } optionsEntityMap    If provided, it is expected to be a
+ *   map of modelName fields to `label` and `value` keys used by
+ *   `mapOptionsCallback`.
+ */
+export class ModelSelect extends Component {
+	static propTypes = {
+		selectConfiguration: PropTypes.shape( {
+			...REACT_SELECT_TYPES,
+		} ),
+		modelEntities: PropTypes.array,
+		modelName: PropTypes.oneOf( MODEL_NAMES ),
+		mapOptionsCallback: PropTypes.func,
+		optionsEntityMap: PropTypes.object,
+		queryData: PropTypes.shape( {
+			limit: PropTypes.number,
+			orderBy: PropTypes.string,
+			order: PropTypes.oneOf( [ 'asc', 'desc' ] ),
+		} ),
+		getQueryString: PropTypes.func,
+		selectLabel: PropTypes.string,
+		addAllOptionLabel: PropTypes.string,
+	};
+
+	static defaultProps = {
+		selectConfiguration: {
+			...REACT_SELECT_DEFAULTS,
+			name: uniqueId( 'model-select-' ),
+		},
+		modelEntities: [],
+		modelName: '',
+		mapOptionsCallback: buildOptions,
+		optionsEntityMap: null,
+		queryData: {
+			limit: 100,
+			order: 'desc',
+		},
+		selectLabel: '',
+		addAllOptionLabel: '',
+	};
+
+	static getDerivedStateFromProps( props ) {
+		const { selectConfiguration } = props;
+		const options = ModelSelect.getOptions( props );
+		const updated = {
+			options,
+			value: ModelSelect.getOptionObjectForValue(
+				selectConfiguration.defaultValue, options
+			),
+		};
+		return {
+			...REACT_SELECT_DEFAULTS,
+			...selectConfiguration,
+			...updated,
+		};
+	}
```

Comment:
I originally intended to link to the "You probably don't need derived state" in my original description for this pull but linked to a different (but still related) post by mistake.  For context, here's what I wrote in the description for the pull:

> I think I may need to tweak how I'm using derivedStateFromProps in ModelSelect yet. Basically I should only be changing state if there is a change detected for the incoming props with the current state (see example here). I can probably utilize memoize a bit in the some of the helper functions as well. This can be something done after this pull is merged and we are using it in blocks (so we can measure improvements).

We talked about this in slack but to summarize:

- I'm aware of the potential issues.
- I'm not convinced (yet) that we're doing things wrong (or in a way that is problematic). 
- I fully intend on investigating further to see if we can do what's needed without using `getDerivedStateFromProps` but think this is something that will be easier to investigate in actual implementations of this component.  So I'd prefer to get this merged in then create another pull to test with and try different potential solutions.